### ISSUE TITLE
Slug updater doesn't consider records with same name

### ISSUE THREAD

**Issue by h-e-l-l-o-w-o-r-l-d** [cobwebch/external_import#195]
Hi there,

I think I found a bug. I need to import records that partially have same titles (same events but different dates). This is usually no problem as the TYPO3  behaviour is to add a prefix to the slug where a slug already exists -> -1, -2 and so on.

When I insert records, external import respects existing slugs and adds a prefix independend of the value in updateSlugs. On update operations though, there won't be any prefixes added when updateSlugs = true. All records with the same name gets the same slug. This will lead to records that are not reachable by an URL.

Here is the behaviour:

insert and updateSlugs = false -> okay
insert and updateSlugs = true -> okay
update and updateSlugs = false -> okay because nothing happens with the slugs anyway
update and updateSlugs = true -> not okay because of missing prefixes

Hint: derhansen/sf_event_mgt had an slug updater before version 5.0 that could be triggered via CLI and worked well. https://github.com/derhansen/sf_event_mgt/issues/731

Can you do anything about it?

**Comment by fsuter** [comment]
There's indeed a difference, since for insert operations, the slug generation is triggered by the TYPO3 Core Engine, whereas for updates I needed to add my own processing inside External Import.

I'll need to create a test scenario and check what happens. I created a slug utility class which wraps around `\TYPO3\CMS\Core\DataHandling\SlugHelper` provided by the TYPO3 Core, but maybe there's more to its usage than I considered.

Thanks for the pointer!

**Comment by h-e-l-l-o-w-o-r-l-d** [comment]
Thanks for looking into it! :)

**Comment by fsuter** [comment]
I found the issue. The Core class is fine, the bug was in my utility class: I had designed the code to generate all the slugs first and save them to the database afterwards. This meant that if 2 slugs were the same it was not detected. I changed the code to save each slug as soon at it is generated, so that duplicates can be correctly detected by the Core.

**Comment by fsuter** [comment]
Question: what version of TYPO3 are you using? I have planned to release this bug fix with the new major version of External Import, which will drop support with v9. Is that fine with you? If not, I'm willing to make the effort to have a bugfix release on branch 5.1.

**Comment by h-e-l-l-o-w-o-r-l-d** [comment]
Great you found it and thank you for looking into it! We use TYPO3 10.4.21. Of course a fast fix for the current branch would be nice (i suppose 6.x will take some time). But no stress, i can just patch it.

**Comment by fsuter** [comment]
Actually, I plan to release very soon, following close on the release of TYPO3 11.5.0 :-)

**Comment by h-e-l-l-o-w-o-r-l-d** [comment]
Sorry to dissapoint you but it's still not working. :-( Updated to major version, cleared all caches and tried to import records with same title. It's exact the same (wrong) behaviour like in my initial description: same slugs for different records are stored (aka not prefixed) and thus records not reachable in the frontend.

Can I help further to solve this?

TYPO3 10.4.21
cobweb/external_import 6.0.1
cobweb/svconnector_json 3.0
(cobweb/svconnector 4.0)

**Comment by fsuter** [comment]
Well, I'm sorry for you. If you could debug more finely it would be very helpful. Otherwise I will look into that when I have time, but the days ahead are packed.

**Comment by h-e-l-l-o-w-o-r-l-d** [comment]
"debug more finely"... what could that be? I'm glad to help but I don't know how to describe it better or what kind of tests/debug steps are needed. I could record a video with insights about source records, the import step and database change after the import. Would that be useful?

In an earlier post you said you found the issue... Did you just changed your code or did you test that with an import?

**Comment by fsuter** [comment]
I haven't added unit or functional tests. I'm pretty sure I did some manual testing, but I can't remember precisely now.

Debugging more finely would be to follow the code while it is running with a debugger and try to understand why existing identical slugs are not detected (and that would happen in Core class `\TYPO3\CMS\Core\DataHandling\SlugHelper`). Debugging would also allow you to check the database after each record is handled and verify that the newly generated slug is indeed saved. I understand that this quite heavy, but it's how I would go about it.

**Comment by h-e-l-l-o-w-o-r-l-d** [comment]
I can confirm that all affected records (where field "title" changed) got a new slug.

As for the debugging procedure I'm afraid to not have possibilities (and maybe know how) to do that. Video is not useful for a start? Only PM of course.

**Comment by fsuter** [comment]
Let's try. Send me a video by PM.

**Comment by h-e-l-l-o-w-o-r-l-d** [comment]
Done and sent to your gmail address.

**Comment by fsuter** [comment]
Well received. I'll take a look as soon as I can.

**Comment by fsuter** [comment]
The "uniqueInTable" condition was not handled.

**Comment by h-e-l-l-o-w-o-r-l-d** [comment]
Hi,

thank for the patch. I installed it successfully but unfortunately it doesn't solve the problem.

But I have something interesting to report: I had the idea to change the EXT:sf_event_mgt setting "records.slugBehaviour" from unique to uniqueInSite and with this setting it works! See:

https://docs.typo3.org/p/derhansen/sf_event_mgt/6.0/en-us/ForAdministrators/Configuration/Extension/Index.html?highlight=uniqueinsite

Unfortunately uniqueInSite it not what we can use, because we have records that are shared between sites. See manual above.

BTW, EXT:news has the exact same setting, but it's not documented in the manual. I reckon the faulty behavior would be the same with EXT:news.

If that finding doesn't help, I recommend to contact the author Torben Hansen who usually responds very fast and seems to be a nice guy. ;o) Maybe he has a clue what's going on. Or if you want, I can point him to this issue!?

**Comment by fsuter** [comment]
I took a second look and it's my mistake. The "unique" property described in the TCA reference is what corresponds to the method called `isUniqueInTable` in the Core. I got confused by this double naming. I'll fix my patch later.

**Comment by fsuter** [comment]
I did right away after all ;-) Can you try again? The previous patch is modified by: https://github.com/cobwebch/external_import/commit/516f6dfd6da65a008554ab18f5f7155b97d74295

**Comment by h-e-l-l-o-w-o-r-l-d** [comment]
Good job, now it works! \o/ Thank you so much!

One question: If I want to have updateSlug on a record basis, or in other words, if I only want to have the updateSlug mechanism for certain records... then I would have to separate them and create a dedicated process with updateSlug  = true and the others with false, right? Or are there any other possibilities with custom steps or the like where i can check some things? The reason is that I actually only want to change the slug if the title changed, so the URL remains valid. With updateSlug = true all slugs get changed, so non-enumerative appendixes become new numbers like this:

slug-1 -> slug-1
slug-3 -> slug-2
slug-5 -> slug 3

Thanks again! Is 6.0.2 around the corner or should I stay with the patch for now?

**Comment by fsuter** [comment]
Huh, I don't really know what would be best. I hadn't thought about this renumbering happening, but it is logical given the code I wrote :-) I can't really see a simple way because you would need to intervene at some point where you have both the old and new title in order to compare them. A possibility could be to add a custom step (between `TransformDataStep` and `StoreDataStep`), use it to load the old title into the `extraData` array of the `Data` object. Then you could use this information later to generate the slugs yourself (so you would set `updateSlug` to `false`) and use the existing code of `SlugUtility` inside an event or a custom step. Just thinking of the top of my head, it would definitely need some more digging.

Please keep the patch, I don't know when I will have time to release, but since the bug fixes are starting to pile up I may release a 6.0.2 rather soon (I am hesitant to put a bit more stuff in the next release and make it 6.1.0). Anyway, not today ;-)

**Comment by h-e-l-l-o-w-o-r-l-d** [comment]
Thanks for the hint. I implemented it with a custom step, just a bit different. The information if a title changed was already in the source data, so right after StoreDataStep (when new titles are saved) I checked the rawData and compared with processed records if the record is updated at all. If yes, store uid in an array. At the end trigger the SlugUtility on those uids. This works!

Happy. =)
